File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
  """
  Assume:
  1 ---- 2         1: [2, 4]
  |      |         2: [1, 3]  
  |      |         3: [2, 4]
  4 ---- 3         4: [1, 3]
  
  =========================================================================================================
  Calling dfsClone(node):
   
  Call stack                    OC_map updates                                
  clone(1) ->                   1 => clone(1)                                 
    clone(2) ->                 2 => clone(2)                                  
      clone(1) ->               -                                             
      clone(3) ->               3 => clone(3)                                 
        clone(2) ->             -                      
        clone(4) ->             4 = > clone(4)
          clone(1) ->           -
          clone(3) ->.          -
  """
  def __init__(self):
    self.OC_map = {} # an original => clone maping

  def cloneGraph(self, node: 'Node') -> 'Node':
    if node is None: return None

    # if node is not cloned yet, clone the node 
    if node.val not in self.OC_map:
      clone = Node(node.val) 
      self.OC_map[node.val] = clone 
      for nei in node.neighbors:
        clone.neighbors.append(self.cloneGraph(nei))

    return self.OC_map[node.val]

      

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/


class Solution {
public:
    // original => clone
    unordered_map<int, Node*> mp; // declaring map, to check whwther we have a copy of node or not and also to store copy
    
    Node* cloneGraph(Node* node) {
        if(node == NULL) // if node is null, then simply return null
        {
            return NULL;
        }
        
        // for a node, we will check whether we already creates a copy of this or not. (Note that we can use the node's val as key of the map as they are guarenteed to be unique across the graph)
        // If it is present in map that means we already created a copy of this, so simply return it 
      
        // But if not present in map, we need to first clone the node, then estabish all the edges with its neighbors 
        
        if(mp.find(node->val) == mp.end()) // if not present in map
        {
            mp[node->val] = new Node(node -> val, {}); // make a copy
            
            for(auto adj: node -> neighbors) // travel in adjcant
            {
                mp[node->val] -> neighbors.push_back(cloneGraph(adj)); //add copy
            }
        }
        
        // either node is already cloned  \
        // or it's newly cloned          -- the node is cloned and will be returned to its caller
        return mp[node->val]; 
    }
};